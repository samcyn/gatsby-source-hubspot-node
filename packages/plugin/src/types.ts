import type { PluginOptions as GatsbyDefaultPluginOptions, IPluginRefOptions, SourceNodesArgs } from 'gatsby';
import { RequestInit } from 'node-fetch';

export interface IPostImageInput {
  url: string;
  alt: string;
  width: number;
  height: number;
}

interface IBlogAuthor {
  avatar?: string;
  id?: string;
}

interface IBlogMeta {
  tag_ids?: Array<string>;
  topic_ids?: Array<string>;
}

// mirror this after hubspot blog post shape
export interface IPostInput {
  id: string;
  absolute_url?: string;
  author?: string;
  author_name?: string;
  author_username?: string;
  blog_author?: IBlogAuthor;
  blog_author_id?: string;
  category?: number;
  category_id?: string;
  created?: number;
  current_state?: string;
  currently_published?: boolean;
  featured_image?: string;
  featured_image_alt_text?: string;
  featured_image_height?: number;
  featured_image_length?: number;
  featured_image_width?: number;
  html_title?: string;
  is_published?: boolean;
  keywords?: Array<string>;
  label?: string;
  language?: string;
  meta?: IBlogMeta;
  meta_description?: string;
  name?: string;
  page_title?: string;
  portal_id?: number;
  post_body?: string;
  post_summary?: string;
  publish_date?: number;
  publish_date_local_time?: number;
  publish_immediately?: boolean;
  published_at?: number;
  published_by_email?: string;
  published_by_id?: number;
  published_by_name?: string;
  published_url?: string;
  slug: string;
}

export type NodeBuilderInput<T> = { type: string; data: T | Record<string, unknown> };

interface IPluginOptionsKeys<T> {
  endpoint: string;
  nodeType?: string;
  /**
   * document in node fetch
   * @params RequestInit
   * @see docs available here https://github.com/node-fetch/node-fetch
   */
  requestOptions?: RequestInit;
  searchParams?: Record<string, string>;
  schemaCustomizationString?: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  apiResponseFormatter?: (args: any) =>
    | Array<T>
    | Array<{
        [x: string]: unknown;
      }>;
  nodeBuilderFormatter?: ({
    gatsbyApi,
    input,
    pluginOptions,
  }: {
    gatsbyApi: SourceNodesArgs;
    input: NodeBuilderInput<T>;
    pluginOptions?: IPluginOptionsInternal<T>;
  }) => void;
}

/**
 * Gatsby expects the plugin options to be of type "PluginOptions" for gatsby-node APIs (e.g. sourceNodes)
 */
export interface IPluginOptionsInternal<T = IPostInput> extends IPluginOptionsKeys<T>, GatsbyDefaultPluginOptions {}

/**
 * These are the public TypeScript types for consumption in gatsby-config
 */
export interface IPluginOptions<T = IPostInput> extends IPluginOptionsKeys<T>, IPluginRefOptions {}
