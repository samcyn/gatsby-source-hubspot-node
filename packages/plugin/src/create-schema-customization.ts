import type { GatsbyNode } from 'gatsby';
import { NODE_TYPES } from './constants';

/**
 * By default Gatsby, infers the data types for each node. This can be sometimes brittle or lead to hard-to-debug errors.
 * However, you as the source plugin author should explicitly define the data types. This way you override the inference.
 * This has multiple benefits:
 * - All data types are correct and consistent across all nodes
 * - Users can have optional data/missing data. With inference this would break builds, with explicit data types it will work
 * - You can set up relationships between nodes
 *
 * Important: For Valhalla, you need to explicitly define the data types for all nodes.
 * @see https://www.gatsbyjs.com/docs/reference/graphql-data-layer/schema-customization/#explicitly-defining-data-types
 * @see https://www.gatsbyjs.com/docs/reference/config-files/gatsby-node/#createSchemaCustomization
 */
export const createSchemaCustomization: GatsbyNode[`createSchemaCustomization`] = ({ actions }) => {
  const { createTypes } = actions;

  /**
   * Two things are happening here:
   * - The `Post` and `Author` types are being explicitly defined with all their fields
   * - The `author` field on the `Post` type is being linked to the `Author` type via a foreign-key relationship
   * @see https://www.gatsbyjs.com/docs/reference/graphql-data-layer/schema-customization/#foreign-key-fields
   */
  createTypes(`
    type BlogAuthor {
      avatar: String
      id: ID
    }
    
    type Meta {
      tag_ids: [ID]
      topic_ids: [ID]
    }  

    type ${NODE_TYPES.Post} implements Node {
      id: ID!
      absolute_url: String
      author: String
      author_name: String
      author_username: String
      blog_author: BlogAuthor
      blog_author_id: ID
      category: Int
      category_id: ID
      created: Int
      current_state: String
      currently_published: Boolean
      featured_image: ${NODE_TYPES.Asset} @link
      featured_image_alt_text: String
      featured_image_height: Int
      featured_image_length: Int
      featured_image_width: Int
      html_title: String
      is_published: Boolean
      keywords: [String]
      label: String
      language: String
      meta: Meta
      meta_description: String
      name: String
      page_title: String
      portal_id: Int
      post_body: String
      post_summary: String
      publish_date: Int
      publish_date_local_time: Int
      publish_immediately: Boolean
      published_at: Int
      published_by_email: String
      published_by_id: Int
      published_by_name: String
      published_url: String
      slug: String!
    }

    type ${NODE_TYPES.Asset} implements Node & RemoteFile {
      url: String!
      alt: String!
      width: Int!
      height: Int!
    }
  `);

  /**
   * You most often will use SDL syntax to define your data types. However, you can also use type builders for more advanced use cases
   * @see https://www.gatsbyjs.com/docs/reference/graphql-data-layer/schema-customization/#gatsby-type-builders
   */
};
