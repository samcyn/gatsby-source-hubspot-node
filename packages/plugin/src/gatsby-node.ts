export const helloWord = () => {
  console.log('1233');
};
