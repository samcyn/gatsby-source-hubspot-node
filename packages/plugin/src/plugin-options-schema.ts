import type { GatsbyNode } from 'gatsby';
import type { ObjectSchema } from 'gatsby-plugin-utils';

import { IPluginOptionsInternal } from './types';

/**
 * When you expose options for your plugin, it's best practice to validate the user input.
 * You can use the pluginOptionsSchema API to do this, which is powered by Joi: https://joi.dev/
 * If for example a user would forget to add the endpoint option, Gatsby will show a validation error.
 * @see https://www.gatsbyjs.com/docs/reference/config-files/gatsby-node/#pluginOptionsSchema
 * @see https://www.gatsbyjs.com/docs/how-to/plugins-and-themes/configuring-usage-with-plugin-options/
 */
export const pluginOptionsSchema: Exclude<GatsbyNode['pluginOptionsSchema'], undefined> = ({ Joi }): ObjectSchema => {
  return Joi.object<unknown, IPluginOptionsInternal>({
    endpoint: Joi.string().uri().required().description('Hubspot API endpoint'),
    requestOptions: Joi.object<unknown, IPluginOptionsInternal['requestOptions']>(),
    searchParams: Joi.object<unknown, IPluginOptionsInternal['searchParams']>(),
    nodeTypeOptions: Joi.object<unknown, IPluginOptionsInternal['nodeTypeOptions']>({
      nodeType: Joi.string().required().description('node type name'),
      schemaCustomizationString: Joi.string().required(),
      apiResponseFormatter: Joi.func().arity(1).required(),
      nodeBuilderFormatter: Joi.func().arity(1).required(),
    }),
  });
};
