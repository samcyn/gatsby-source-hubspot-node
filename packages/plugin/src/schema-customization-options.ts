export const schemaCustomizationOptions = `
  type BlogAuthor {
    avatar: String
    id: ID
  }

  type Meta {
    tag_ids: [ID]
    topic_ids: [ID]
  }  

  type Post implements Node {
    id: ID!
    absolute_url: String
    author: String
    author_name: String
    author_username: String
    blog_author: BlogAuthor
    blog_author_id: ID
    category: Int
    category_id: ID
    created: Int
    current_state: String
    currently_published: Boolean
    featured_image: Asset @link
    featured_image_alt_text: String
    featured_image_height: Int
    featured_image_length: Int
    featured_image_width: Int
    html_title: String
    is_published: Boolean
    keywords: [String]
    label: String
    language: String
    meta: Meta
    meta_description: String
    name: String
    page_title: String
    portal_id: Int
    post_body: String
    post_summary: String
    publish_date: Int
    publish_date_local_time: Int
    publish_immediately: Boolean
    published_at: Int
    published_by_email: String
    published_by_id: Int
    published_by_name: String
    published_url: String
    slug: String!
  }

  type Asset implements Node & RemoteFile {
    url: String!
    alt: String!
    width: Int!
    height: Int!
  }
`;
