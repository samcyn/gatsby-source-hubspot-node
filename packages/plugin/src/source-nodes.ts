import type { GatsbyNode } from 'gatsby';

import type { IPluginOptionsInternal } from './types';
import { CACHE_KEYS, ERROR_CODES } from './constants';
import {
  apiResponseFormatter as defaultApiResponseFormatter,
  fetchRequest,
  nodeBuilderFormatter as defaultNodeBuilderFormatter,
} from './utils';

let isFirstSource = true;

/**
 * The sourceNodes API is the heart of a Gatsby source plugin. This is where data is ingested and transformed into Gatsby's data layer.
 * @see https://www.gatsbyjs.com/docs/reference/config-files/gatsby-node/#sourceNodes
 */
export const sourceNodes: GatsbyNode['sourceNodes'] = async (gatsbyApi, pluginOptions: IPluginOptionsInternal) => {
  const { actions, reporter, cache, getNodes } = gatsbyApi;
  const { touchNode } = actions;
  const {
    endpoint,
    requestOptions,
    searchParams,
    nodeType = 'Post',
    apiResponseFormatter,
    nodeBuilderFormatter,
  } = pluginOptions;

  /**
   * It's good practice to give your users some feedback on progress and status. Instead of printing individual lines, use the activityTimer API.
   * This will give your users a nice progress bar and can you give updates with the .setStatus API.
   * In the end your users will also have the exact time it took to source the data.
   * @see https://www.gatsbyjs.com/docs/reference/config-files/node-api-helpers/#reporter
   */
  const sourcingTimer = reporter.activityTimer(`Sourcing from plugin API`);
  sourcingTimer.start();

  if (isFirstSource) {
    /**
     * getNodes() returns all nodes in Gatsby's data layer
     */
    getNodes().forEach((node) => {
      /**
       * "owner" is the name of your plugin, the "name" you defined in the package.json
       */
      if (node.internal.owner !== 'plugin') {
        return;
      }

      /**
       * Gatsby aggressively garbage collects nodes between runs. This means that nodes that were created in the previous run but are not created in the current run will be deleted. You can tell Gatsby to keep old, but still valid nodes around, by "touching" them.
       * For this you need to use the touchNode API.
       *
       * However, Gatsby only checks if a node has been touched on the first sourcing. This is what the "isFirstSource" variable is for.
       * @see https://www.gatsbyjs.com/docs/reference/config-files/actions/#touchNode
       */
      touchNode(node);
    });

    isFirstSource = false;
  }

  /**
   * If your API supports delta updates via e.g. a timestamp or token, you can store that information via the cache API.
   *
   * The cache API is a key-value store that persists between runs.
   * You should also use it to persist results of time/memory/cpu intensive tasks.
   * @see https://www.gatsbyjs.com/docs/reference/config-files/node-api-helpers/#cache
   */
  const lastFetchedDate: number = await cache.get(CACHE_KEYS.Timestamp);
  const lastFetchedDateCurrent = Date.now();

  /**
   * The reporter API has a couple of methods:
   * - info: Print a message to the console
   * - warn: Print a warning message to the console
   * - error: Print an error message to the console
   * - panic: Print an error message to the console and exit the process
   * - panicOnBuild: Print an error message to the console and exit the process (only during "gatsby build")
   * - verbose: Print a message to the console that is only visible when the "verbose" flag is enabled (e.g. gatsby build --verbose)
   * @see https://www.gatsbyjs.com/docs/reference/config-files/node-api-helpers/#reporter
   *
   * Try to keep the terminal information concise and informative. You can use the "verbose" method to print more detailed information.
   * You don't need to print out every bit of detail your plugin is doing as otherwise it'll flood the user's terminal.
   */
  reporter.verbose(`[plugin] Last fetched date: ${lastFetchedDate}`);

  /**
   * Fetch data from the example API. This will differ from your implementation and personal preferences on e.g. which library to use.
   * A good general recommendation is: https://github.com/sindresorhus/got
   */
  try {
    const response = await fetchRequest({ endpoint, requestOptions, searchParams });

    if ('status' in response && response.status === 'error') {
      const { correlationId, errorType, message } = response;

      sourcingTimer.panicOnBuild({
        id: ERROR_CODES.HubspotSourcing,
        error: Error(message),
        context: {
          sourceMessage: errorType,
          errorCode: 500,
        },
        correlationId,
      });
      return;
    }
    /**
     * Gatsby's cache API uses LMDB to store data inside the .cache/caches folder.
     *
     * As mentioned above, cache the timestamp of last sourcing.
     * The cache API accepts "simple" data structures like strings, integers, arrays.
     * For example, passing a Set or Map won't work because the "structuredClone" option is purposefully not enabled:
     * https://github.com/kriszyp/lmdb-js#serialization-options
     */
    await cache.set(CACHE_KEYS.Timestamp, lastFetchedDateCurrent);

    const postFormatter = apiResponseFormatter ? apiResponseFormatter : defaultApiResponseFormatter;

    const posts = postFormatter(response);

    /**
     * Up until now the terminal output only showed "Sourcing from plugin API" and a timer. Via the "setStatus" method you can add more information to the output.
     * It'll then print "Sourcing from plugin API - Processing X posts and X authors"
     */
    sourcingTimer.setStatus(`Processing ${posts.length} posts`);

    /**
     * Iterate over the data and create nodes
     */
    const postNodeBuilderFormatter = nodeBuilderFormatter ? nodeBuilderFormatter : defaultNodeBuilderFormatter;

    for (const post of posts) {
      postNodeBuilderFormatter({ gatsbyApi, input: { type: nodeType, data: post } });
    }

    sourcingTimer.end();
  } catch (error) {
    sourcingTimer.panicOnBuild({
      id: ERROR_CODES.HubspotSourcing,
      error,
      context: {
        sourceMessage: 'Sourcing from the Hubspot API failed',
        errorCode: 500,
      },
    });
  }
};
