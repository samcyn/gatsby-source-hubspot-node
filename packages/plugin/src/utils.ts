import type { NodeInput, SourceNodesArgs } from 'gatsby';
import fetch, { HeadersInit } from 'node-fetch';

import { IPluginOptionsInternal, IPostImageInput, IPostInput } from './types';
import { NODE_TYPES } from './constants';
import { IRemoteImageNodeInput } from 'gatsby-plugin-utils/index';

const defaultHeaders = {
  'Content-Type': 'application/json',
} satisfies HeadersInit;

interface IApiResponse<T> {
  correlationId: string;
  errorType: string;
  message: string;
  status: string;
  [x: string]: Array<T> | unknown;
}

export async function fetchRequest<T = IPostInput>(
  pluginOptions: Pick<IPluginOptionsInternal<T>, 'endpoint' | 'requestOptions' | 'searchParams'>
): Promise<IApiResponse<T>> {
  const { endpoint, requestOptions, searchParams } = pluginOptions;

  let url = endpoint;

  if (searchParams) {
    const params = new URLSearchParams(searchParams);
    // Check if endpoint already contains query parameters
    const hasQueryParams = endpoint.includes('?');
    url += hasQueryParams ? `&${params.toString()}` : `?${params.toString()}`;
  }

  const { headers, method, ...rest } = requestOptions;

  const response = await fetch(url, {
    method: method || 'GET',
    headers: {
      ...defaultHeaders,
      ...headers,
    },
    ...rest,
  });

  const result = await response.json();
  return result;
}

export const apiResponseFormatter: IPluginOptionsInternal<IPostInput>['nodeTypeOptions']['apiResponseFormatter'] = (
  response
) => {
  return response.objects;
};

export const nodeBuilderFormatter: IPluginOptionsInternal<IPostInput>['nodeTypeOptions']['nodeBuilderFormatter'] = ({
  gatsbyApi,
  input,
}) => {
  const id = gatsbyApi.createNodeId(`${input.type}-${input.data.id}`);

  const extraData: Record<string, unknown> = {};

  if (input.type === 'Post') {
    // eslint-disable-next-line @typescript-eslint/naming-convention
    const { featured_image, featured_image_alt_text, featured_image_height, featured_image_width } =
      input.data as IPostInput;

    const featuredImage = {
      url: featured_image,
      alt: featured_image_alt_text,
      width: featured_image_width,
      height: featured_image_height,
    } satisfies IPostImageInput;

    const assetId = createAssetNode(gatsbyApi, featuredImage);

    // This sets the autogenerated Node ID onto the "image" key of the Post node. Then the @link directive in the schema will work.
    extraData.featured_image = assetId;
  }

  const node = {
    ...input.data,
    ...extraData,
    id,
    parent: null,
    children: [],
    internal: {
      type: input.type,
      /**
       * The content digest is a hash of the entire node.
       * Gatsby uses this internally to determine if the node needs to be updated.
       */
      contentDigest: gatsbyApi.createContentDigest(input.data),
    },
  } satisfies NodeInput;

  /**
   * Add the node to Gatsby's data layer. This is the most important piece of a Gatsby source plugin.
   * @see https://www.gatsbyjs.com/docs/reference/config-files/actions/#createNode
   */
  gatsbyApi.actions.createNode(node);
};

export function createAssetNode(gatsbyApi: SourceNodesArgs, data: IPostImageInput) {
  const id = gatsbyApi.createNodeId(`${NODE_TYPES.Asset}-${data.url}`);

  /**
   * For Image CDN and the "RemoteFile" interface, these fields are required:
   * - url
   * - filename
   * - mimeType
   * For images, these fields are also required:
   * - width
   * - height
   */
  const assetNode = {
    id,
    url: data.url,
    mimeType: `image/jpg`,
    filename: data.url,
    width: data.width,
    height: data.height,
    placeholderUrl: `${data.url}&w=%width%&h=%height%`,
    alt: data.alt,
    parent: null,
    children: [],
    internal: {
      type: NODE_TYPES.Asset,
      contentDigest: gatsbyApi.createContentDigest(data),
    },
  } satisfies IRemoteImageNodeInput;

  gatsbyApi.actions.createNode(assetNode);

  /**
   * Return the id so it can be used for the foreign key relationship on the Post node.
   */
  return id;
}
