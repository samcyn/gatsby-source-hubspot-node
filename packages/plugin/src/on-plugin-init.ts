import type { GatsbyNode } from 'gatsby';
import { ERROR_CODES } from './constants';
import { wrapWithAsterisks } from './utils';

/**
 * Use the onPluginInit API to set up things that should be run before the plugin is initialized.
 * You can use this e.g. for the reporter.setErrorMap API. Using this API is optional, but highly recommended.
 * By providing your own errors you can make it clearer to your users where the error is coming from and you can control what to show.
 * Using the ERROR_CODES enum you can reference the IDs without knowing them by hard.
 * @see https://www.gatsbyjs.com/docs/reference/config-files/gatsby-node/#onPluginInit
 * @see https://www.gatsbyjs.com/docs/reference/config-files/node-api-helpers/#reporter
 */
export const onPluginInit: GatsbyNode['onPluginInit'] = ({ reporter }) => {
  const info = `Hey, thanks for considering installing "gatsby-source-hubspot-node",
  for documentation see https://github.com/samcyn/gatsby-source-hubspot-node`;

  reporter.info(wrapWithAsterisks(info));
  reporter.setErrorMap({
    [ERROR_CODES.HubspotSourcing]: {
      text: (context) => `${context.sourceMessage}: ${context.errorCode}`,
      level: 'ERROR',
      type: 'API.NODE.EXECUTION',
      category: 'THIRD_PARTY',
      docsUrl: 'https://legacydocs.hubspot.com/docs',
    },
  });
};
