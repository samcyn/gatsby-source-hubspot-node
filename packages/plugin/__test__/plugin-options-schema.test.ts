import { expect, describe, it } from 'vitest';
import { testPluginOptionsSchema } from 'gatsby-plugin-utils';

import { pluginOptionsSchema } from '../../plugin/src/plugin-options-schema';

describe(`pluginOptionsSchema`, () => {
  it(`should match snapshot of pluginOptionsSchema`, () => {
    expect(pluginOptionsSchema).toMatchSnapshot();
  });

  it(`should invalidate incorrect options`, async () => {
    const options = {
      endpoint: undefined,
    };

    const { isValid, errors } = await testPluginOptionsSchema(pluginOptionsSchema, options);

    expect(isValid).toBe(false);
    expect(errors).toEqual([`"endpoint" is required`]);
  });
  it(`should invalidate incorrect endpoint string`, async () => {
    const options = {
      endpoint: `foo`,
    };

    const { isValid, errors } = await testPluginOptionsSchema(pluginOptionsSchema, options);

    expect(isValid).toBe(false);
    expect(errors).toEqual([`"endpoint" must be a valid uri`]);
  });
  it(`should validate correct options`, async () => {
    const options = {
      endpoint: `http://localhost:4000/graphql`,
    };

    const { isValid, errors } = await testPluginOptionsSchema(pluginOptionsSchema, options);

    expect(isValid).toBe(true);
    expect(errors).toEqual([]);
  });
  it(`should invalidate incorrect nodeTypeOptions option when one or more missing required fields are not provided`, async () => {
    const options = {
      endpoint: 'https://samsoniyanda.com',
      nodeTypeOptions: {
        nodeType: 'Post',
      },
    };

    const { isValid, errors } = await testPluginOptionsSchema(pluginOptionsSchema, options);

    expect(isValid).toBe(false);
    expect(errors).toEqual([
      `"nodeTypeOptions.schemaCustomizationString" is required`,
      `"nodeTypeOptions.apiResponseFormatter" is required`,
      `"nodeTypeOptions.nodeBuilderFormatter" is required`,
    ]);
  });
  it(`should validate correct nodeTypeOptions option when required fields are provided`, async () => {
    const options = {
      endpoint: 'https://samsoniyanda.com',
      nodeTypeOptions: {
        nodeType: 'Post',
        schemaCustomizationString: `
          type Post implements Node {
            id: ID!
          }
        `,
        apiResponseFormatter(response: unknown) {
          return response;
        },
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        nodeBuilderFormatter(_node: unknown) {},
      },
    };

    const { isValid, errors } = await testPluginOptionsSchema(pluginOptionsSchema, options);

    expect(isValid).toBe(true);
    expect(errors).toEqual([]);
  });
});
