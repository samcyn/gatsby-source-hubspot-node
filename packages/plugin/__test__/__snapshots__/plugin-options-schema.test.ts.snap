// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`pluginOptionsSchema > should match snapshot of pluginOptionsSchema 1`] = `[Function]`;
